### ISSUE TITLE
No longer opens on active monitor (but on the monitor it shows up the first time) [Unity]

### ISSUE THREAD

**Issue by hotice** [kupferlauncher/kupfer#45]
Kupfer no longer opens on the active monitor in my dual-monitor setup (Ubuntu 16.10; update: it occurs under Unity but not in GNOME Shell), but on the monitor it shows up the first time. I missed this bug somehow, and I'm not sure when it started, but it occurs for me with both Kupfer v308.1 and v307.

Here's the debug, though it doesn't seem to give any hints about this:
```
$ kupfer --debug
D [kupfer.main]: Version: kupfer v308.1
D [kupfer.scheduler] Scheduler: Loading
D [kupfer.core.plugins]: Loading kupfer.plugin.applications
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/applications.py
D [kupfer.core.sources]: Add <kupfer.plugin.applications.OpenWith>
D [kupfer.core.sources]: Add <kupfer.plugin.applications.SetDefaultApplication>
D [kupfer.core.sources] SourcePickler: Loading Applications from k3aa790ae0ab77a9888a349a397a02e4b-v5.pickle.gz
D [kupfer.core.sources]: Add <kupfer.plugin.applications.AppSource>
D [kupfer.core.plugins]: Loading kupfer.plugin.archivemanager
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/archivemanager.py
D [kupfer.core.sources]: Add <kupfer.plugin.archivemanager.UnpackHere>
D [kupfer.core.sources]: Add <kupfer.plugin.archivemanager.CreateArchive>
D [kupfer.core.sources]: Add <kupfer.plugin.archivemanager.CreateArchiveIn>
D [kupfer.core.plugins]: Loading kupfer.plugin.audacious
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/audacious.py
D [kupfer.core.sources]: Add <class 'kupfer.plugin.audacious.AudaciousSource'>
D [kupfer.core.sources] SourcePickler: Loading Audacious from ka110115853ddd35bc2715eb83a190557-v5.pickle.gz
D [kupfer.core.sources]: Add <kupfer.plugin.audacious.AudaciousSource>
D [kupfer.core.plugins]: Loading kupfer.plugin.calculator
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/calculator.py
D [kupfer.core.sources]: Add <kupfer.plugin.calculator.Calculate>
D [kupfer.core.plugins]: Loading kupfer.plugin.clipboard
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/clipboard.py
D [kupfer.core.sources]: Add <kupfer.plugin.clipboard.ClearClipboards>
D [kupfer.core.sources] SourcePickler: Loading Clipboards from k2c65a3b08d6cd573a4e3127f7df5ab7b-v5.pickle.gz
D [kupfer.core.sources]: Add <kupfer.plugin.clipboard.ClipboardSource>
D [kupfer.core.plugins]: Loading kupfer.plugin.commands
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/commands.py
D [kupfer.core.sources]: Add <kupfer.plugin.commands.CommandTextSource>
D [kupfer.core.sources]: Add <kupfer.plugin.commands.PassToCommand>
D [kupfer.core.sources]: Add <kupfer.plugin.commands.PassToCommand>
D [kupfer.core.sources]: Add <kupfer.plugin.commands.FilterThroughCommand>
D [kupfer.core.sources]: Add <kupfer.plugin.commands.WriteToCommand>
D [kupfer.icons]: Loading icon kupfer-object at 24 for core
D [kupfer.icons]: Loading icon kupfer-object at 128 for core
D [kupfer.icons]: Loading icon kupfer-object-multiple at 24 for core
D [kupfer.icons]: Loading icon kupfer-object-multiple at 128 for core
D [kupfer.core.plugins]: Loading kupfer.plugin.core
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/core/__init__.py
D [kupfer.plugin_support]: Registered alternative icon_renderer: kupfer.plugin.core.gtk
D [kupfer.plugin_support]: Registered alternative terminal: kupfer.plugin.core.gnome-terminal
D [kupfer.plugin_support]: Registered alternative terminal: kupfer.plugin.core.xfce4-terminal
D [kupfer.plugin_support]: Registered alternative terminal: kupfer.plugin.core.lxterminal
D [kupfer.plugin_support]: Registered alternative terminal: kupfer.plugin.core.xterm
D [kupfer.plugin_support]: Registered alternative terminal: kupfer.plugin.core.urxvt
D [kupfer.plugin_support]: Registered alternative terminal: kupfer.plugin.core.konsole
D [kupfer.core.sources]: Add <kupfer.plugin.core.text.BasicTextSource>
D [kupfer.core.sources]: Add <kupfer.plugin.core.text.PathTextSource>
D [kupfer.core.sources]: Add <kupfer.plugin.core.text.URLTextSource>
D [kupfer.core.sources]: Add <kupfer.plugin.core.SearchInside>
D [kupfer.core.sources]: Add <kupfer.plugin.core.CopyToClipboard>
D [kupfer.core.sources]: Add <kupfer.plugin.core.debug.DebugInfo>
D [kupfer.core.sources]: Add <kupfer.plugin.core.debug.Forget>
D [kupfer.core.sources]: Add <kupfer.plugin.core.commands.SaveToFile>
D [kupfer.core.sources]: Add <kupfer.plugin.core.Rescan>
D [kupfer.core.sources]: Add <kupfer.plugin.core.text.OpenTextUrl>
D [kupfer.core.sources]: Add <kupfer.plugin.core.Rescan>
D [kupfer.core.sources]: Add <class 'kupfer.plugin.core.contents.KupferSource'>
D [kupfer.core.sources]: Add <class 'kupfer.plugin.core.debug.ComposedSource'>
D [kupfer.core.sources]: Add <kupfer.plugin.core.internal.CommandResults>
D [kupfer.core.sources]: Add <kupfer.plugin.core.contents.KupferSource>
D [kupfer.core.sources]: Add <kupfer.plugin.core.internal.KupferInterals>
D [kupfer.core.plugins]: Loading kupfer.plugin.dictionary
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/dictionary.py
D [kupfer.core.sources]: Add <kupfer.plugin.dictionary.LookUp>
D [kupfer.core.plugins]: Loading kupfer.plugin.documents
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/documents.py
D [kupfer.core.sources]: Add <class 'kupfer.plugin.documents.ApplicationRecentsSource'>
D [kupfer.core.sources] SourcePickler: Loading Recent Items from k875b3cebdfd4b9ebcd0396de3828a5fc-v5.pickle.gz
D [kupfer.core.sources]: Add <kupfer.plugin.documents.RecentsSource>
D [kupfer.core.sources] SourcePickler: Loading Places from k3b268c0468d6a50f151376a8d4319726-v5.pickle.gz
D [kupfer.core.sources]: Add <kupfer.plugin.documents.PlacesSource>
D [kupfer.core.plugins]: Loading kupfer.plugin.favorites
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/favorites.py
D [kupfer.core.sources]: Add <kupfer.plugin.favorites.AddFavorite>
D [kupfer.core.sources]: Add <kupfer.plugin.favorites.RemoveFavorite>
D [kupfer.core.sources] SourceDataPickler: Loaded configuration from config-kupfer.plugin.favorites-v2.pickle
D [kupfer.core.sources]: Add <kupfer.plugin.favorites.FavoritesSource>
D [kupfer.core.plugins]: Loading kupfer.plugin.mediaplayers
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/mediaplayers.py
D [kupfer.core.sources]: Add <kupfer.plugin.mediaplayers.Players>
D [kupfer.core.plugins]: Loading kupfer.plugin.notes
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/notes.py
D [kupfer.core.sources]: Add <kupfer.plugin.notes.AppendToNote>
D [kupfer.core.sources]: Add <kupfer.plugin.notes.CreateNote>
D [kupfer.core.sources]: Add <kupfer.plugin.notes.GetNoteSearchResults>
D [kupfer.core.sources]: Add <class 'kupfer.plugin.notes.NotesSource'>
D [kupfer.core.sources] SourcePickler: Loading Notes from kc2341bc9e33cc257352b62a2364163f0-v5.pickle.gz
D [kupfer.core.sources]: Add <kupfer.plugin.notes.NotesSource>
D [kupfer.core.plugins]: Loading kupfer.plugin.session_gnome
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/session_gnome.py
D [kupfer.core.sources]: Add <kupfer.plugin.session_gnome.GnomeItemsSource>
D [kupfer.core.plugins]: Loading kupfer.plugin.show_text
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/show_text.py
D [kupfer.core.sources]: Add <kupfer.plugin.show_text.ShowText>
D [kupfer.core.sources]: Add <kupfer.plugin.show_text.ShowNotification>
D [kupfer.core.plugins]: Loading kupfer.plugin.trash
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/trash.py
D [kupfer.core.sources]: Add <kupfer.plugin.trash.MoveToTrash>
D [kupfer.core.sources] SourcePickler: Loading Trash from k67725c30e6b5980bfcac436b0554cc10-v5.pickle.gz
D [kupfer.core.sources]: Add <kupfer.plugin.trash.TrashSource>
D [kupfer.core.plugins]: Loading kupfer.plugin.triggers
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/triggers.py
D [kupfer.core.sources]: Add <kupfer.plugin.triggers.AddTrigger>
D [kupfer.core.sources] SourceDataPickler: Loaded configuration from config-kupfer.plugin.triggers-v2.pickle
D [kupfer.core.sources]: Add <kupfer.plugin.triggers.Triggers>
D [kupfer.core.plugins]: Loading kupfer.plugin.urlactions
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/urlactions.py
D [kupfer.core.sources]: Add <kupfer.plugin.urlactions.DownloadAndOpen>
D [kupfer.core.sources]: Add <kupfer.plugin.urlactions.DownloadTo>
D [kupfer.core.plugins]: Loading kupfer.plugin.volumes
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/volumes.py
D [kupfer.core.sources]: Add <kupfer.plugin.volumes.VolumesSource>
D [kupfer.core.plugins]: Loading kupfer.plugin.wikipedia
D [kupfer.core.plugins]:   from /usr/share/kupfer/kupfer/plugin/wikipedia.py
D [kupfer.core.sources]: Add <kupfer.plugin.wikipedia.WikipediaSearch>
D [kupfer.core.sources] SourcePickler: Loading andrei from k4b9305cc9f71f1efc7968ecad8f62a30-v5.pickle.gz
D [kupfer.core.sources] SourcePickler: Loading Desktop from k328446105362e9a4b3f663bd88d0f689-v5.pickle.gz
D [kupfer.plugin.triggers] Triggers: Loaded triggers, count: 0
D [kupfer.core.sources] PeriodicRescanner: Registering new campaign, in 10 s
D [kupfer.plugin.mediaplayers] Players: Loaded items
D [kupfer.plugin.volumes] VolumesSource: Loaded items
D [kupfer.plugin.clipboard] ClipboardSource: Loaded items
D [kupfer.obj.sources] SourcesSource: Loaded items
D [kupfer.plugin.triggers] Triggers: Loaded items
D [kupfer.plugin.favorites] FavoritesSource: Loaded items
D [kupfer.core.learn]: Reading from /home/andrei/.config/kupfer/mnemonics.pickle
D [kupfer.scheduler] Scheduler: Loaded
D [kupfer.scheduler] Scheduler: Display
D [kupfer.core.data] DataController: Reloading source root
D [kupfer.ui.browser]: Scale factor 1
D [kupfer.scheduler] Scheduler: After Display
D [kupfer.ui.browser] WindowController: in lazy_setup
D [kupfer.ui.keybindings]: binding '<Ctrl>space'
Kupfer: A free software (GPLv3+) launcher
   Copyright © 2007–2017 Ulrik Sverdrup with others
   https://kupferlauncher.github.io/

[kupfer.ui.browser] WindowController: Trying to register <Ctrl>space to spawn kupfer.. success
D [kupfer.ui.session] SessionClient: Connected to session as client /org/gnome/SessionManager/Client8 
D [kupfer.ui.browser] WindowController: finished lazy_setup
[kupfer.core.sources] PeriodicRescanner: Starting new campaign, interval 3 s
D [kupfer.plugin.mediaplayers] Players: Loaded 0 items
D [kupfer.plugin.triggers] Triggers: Loaded 0 items
D [kupfer.obj.sources] DirectorySource: Loaded 30 items
```

**Comment by bluss** [comment]
Thanks for the report. The first question is if this started all the way back with v300?

**Comment by hotice** [comment]
I think it worked initially, but I can't remember for sure. I'll try to investigate.

**Comment by hotice** [comment]
Well, I just tried v300 it works the same as v308.1, so the bug is there from the beginning of v3, it seems. Hmm what's weird is that I think I would have noticed it before, because I've used Kupfer daily since v302 on a dual monitor setup. It might also be some Ubuntu update that did this. But I may be wrong.

**Comment by bluss** [comment]
If the keybinding works on both monitors, then you at least don't need the multihead plugin (that was not ported to v30x). I really don't want to revive that hack.

**Comment by hotice** [comment]
One more note: Kupfer v208 works on both monitors.

**Comment by bluss** [comment]
Explanation could be this simple: Gtk 3 has the same APIs existing, but they have changed behaviour. Gdk.Screen now always says there is 1 monitor.

https://lazka.github.io/pgi-docs/index.html#Gdk-3.0/classes/Screen.html#Gdk.Screen.get_monitor_at_point

**Comment by bluss** [comment]
Delightfully needs Gtk 3.22 to use the new API..

**Comment by bluss** [comment]
I can't reproduce, for me, the kupfer window shows on the monitor where the mouse pointer is. There are still some apps being launched on the wrong window, but those are separate bugs that need to be tracked down one by one.

**Comment by bluss** [comment]
Switching from enhancement to bug since there is existing monitor-placement code.

**Comment by hotice** [comment]
So it turns out I wasn't crazy. I logged in to GNOME Shell today (Ubuntu 16.10) and Kupfer works properly with two monitors, showing up on the active monitor. I initially used Kupfer v3 in GNOME Shell, that's why I didn't notice this issue. So it looks like it's Unity-specific.

**Comment by bluss** [comment]
Thanks, the clarification is important.

The diversity of window managers and environments is why we need multiple contributors to Kupfer. Especially since Unity is not packaged in other distributions, at least not in Debian at the moment.

So this is my plea to anybody to contribute: Ask me on the irc channel (#kupfer on freenode) if you want to have help get started.

If you used gnome shell, that also means you did not experience bug  #51 ? Just to know if kupfer is totally broken there or not.

**Comment by hotice** [comment]
No, I didn't experience that bug in GNOME Shell 3.20.

**Comment by khurshid-alam** [comment]
@hotice Please drop old python2 patches from ppa, except rst2html patch.

**Comment by bluss** [comment]
I looked through the old debian package patches and they should all have been adressed one way or the other, for example by patches by @KarolBedkowski now integrated, or by fixes in kupfer 3xx

**Comment by hotice** [comment]
The patches are dropped, I just didn't delete them from the debian folder. But they are not used, here's the "series" file which includes the patches that are applied:
```
#uses_fragment.patch
#keyring-3.patch
#help.patch
#tracker-needle.patch
#locales.patch
#dbus.patch
#bookmarks-places.sqlite.patch
use-rst2html-instead-of-rst2html5-for-yakkety.patch
```

By the way, this bug was fixed in the latest Kupfer v313, thank you!

**Comment by bluss** [comment]
Yes! Since I only tested Unity, I hope it didn't re-break gnome shell or other multi monitor setups. For this bug I used a virtualbox with two virtual monitors to run Unity in 16.10. I also saw that the preferences window doesn't always come to the front when opened from the app indicator. It looks like app indicators don't give us an event time, and then it can't be better than that. (So opening preferences is best done by just searching for Kupfer Preferences or using the shortcut which is Ctrl + ; by default).
